### PR TITLE
[MLIR] Finalize whens last

### PR DESCRIPTION
**Pull request by leonardt**
Then when finalize must occur after any passes that modify the circuit

e.g. `insert_coreir_wires` could flatten a value used in the conditional logic, which then requires updating the when builder instance

### CODE REVIEW

File: tests/test_when.py [hunk 1]
```diff
@@ -485,6 +485,7 @@ class _Test(m.Circuit):
     assert check_gold(__file__, f"{basename}.mlir")
 
 
+@pytest.mark.skip("There's a bug in this test")
```
Comment (leonardt) [inline tests/test_when.py:488]: Turns out there's a bug in this test, `x` is defined inside a when, and then referred to in the following otherwise, I'm not sure what the original intention of this was, it might have been a temporary change while I was debugging.  Working on a fix but will put it in a separate PR for clarity.

File: magma/backend/mlir/mlir_compiler.py [hunk 2]
```diff
@@ -39,18 +39,19 @@ def suffix(self):
...
-        # finalize when earlier, but then we may still need this code path as
-        # an option.
         if self.opts.get("flatten_all_tuples", False):
             elaborate_tuples(self.main)
-        finalize_whens(self.main)
 
         insert_coreir_wires(self.main)
         insert_wrap_casts(self.main)
         wire_clocks(self.main)
+
```
Comment (rsetaluri) [inline magma/backend/mlir/mlir_compiler.py:48]: nit: remove white line

File: tests/test_backend/test_mlir/test_mlir_compiler.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import magma as m
+
+
+def test_mlir_compiler_finalize_when_order():
```
Comment (rsetaluri) [inline tests/test_backend/test_mlir/test_mlir_compiler.py:4]: let's move this to an example in `examples.py` as the other ones, maybe named `complex_when_lazy_tuple`

File: tests/test_when.py (hunk 1 above)
Comment (rsetaluri) [inline tests/test_when.py:488]: nit: let's make a more descriptive skip message, maybe `"Test does not pass due to when issue (#1156)"`

File: tests/test_backend/test_mlir/test_mlir_compiler.py (hunk 3 above)
Comment (leonardt) [reply tests/test_backend/test_mlir/test_mlir_compiler.py:5]: I moved it to examples, but it looks like the existing examples flow doesn't use the mlir_compiler interface, so it doesn't test that pass ordering, so I left it in this separate test that uses `m.compile` for now

File: tests/test_backend/test_mlir/test_mlir_compiler.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import magma as m
+from examples import complex_when_lazy_tuple
+
+
+def test_mlir_compiler_finalize_when_order():
+    m.compile("build/complex_when_lazy_tuple", complex_when_lazy_tuple,
+              output="mlir")
```
Comment (rsetaluri) [inline tests/test_backend/test_mlir/test_mlir_compiler.py:7]: nit
```suggestion
    m.compile(
        "build/complex_when_lazy_tuple",
        complex_when_lazy_tuple,
        output="mlir",
    )
```

also let's use a tempdir and kill the build dir and .gitignore

File: tests/test_backend/test_mlir/test_mlir_compiler.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import magma as m
+from examples import complex_when_lazy_tuple
```
Comment (rsetaluri) [inline tests/test_backend/test_mlir/test_mlir_compiler.py:2]: nit
```suggestion

from examples import complex_when_lazy_tuple
```

File: tests/test_backend/test_mlir/test_utils.py [hunk 6]
```diff
@@ -181,7 +181,7 @@ def get_local_examples() -> List[DefineCircuitKind]:
         examples.simple_memory_wrapper,
         examples.simple_undriven_instances,
         examples.simple_neg,
-        examples.simple_array_slice,
+        examples.simple_array_slice
```
Comment (rsetaluri) [inline tests/test_backend/test_mlir/test_utils.py:184]: nit: leave `,`

**Review by rsetaluri — APPROVED** [review]
final comments, looks good if addressed
